src/analysis/normed_space/normed_group_quotient.lean

```diff
@@ -0,0 +1,491 @@
+quotient_add_group.ker_mk  _
+
+/-- The operator norm of the projection is at most `1`. -/
+lemma norm_normed_mk_le (S : add_subgroup M) : ∥S.normed_mk∥ ≤ 1 :=
+normed_group_hom.op_norm_le_bound _ zero_le_one (λ m, by simp [quotient_norm_mk_le])
+
+/-- The operator norm of the projection is `1` if the subspace is not dense. -/
+lemma norm_normed_mk (S : add_subgroup M) (h : (S.topological_closure : set M) ≠ univ) :
+  ∥S.normed_mk∥ = 1 :=
+begin
+  obtain ⟨x, hx⟩ := set.nonempty_compl.2 h,
+  let y := S.normed_mk x,
+  have hy : ∥y∥ ≠ 0,
+  { intro h0,
+    exact set.not_mem_of_mem_compl hx ((quotient_norm_eq_zero_iff S x).1 h0) },
+  refine le_antisymm (norm_normed_mk_le S) (le_of_forall_pos_le_add (λ ε hε, _)),
+  suffices : 1 ≤ ∥S.normed_mk∥ + min ε ((1 : ℝ)/2),
+  { exact le_add_of_le_add_left this (min_le_left ε ((1 : ℝ)/2)) },
+  have hδ := sub_pos.mpr (lt_of_le_of_lt (min_le_right ε ((1 : ℝ)/2)) one_half_lt_one),
+  have hδpos : 0 < min ε ((1 : ℝ)/2) := lt_min hε one_half_pos,
+  have hδnorm := mul_pos (div_pos hδpos hδ) (lt_of_le_of_ne (norm_nonneg y) hy.symm),
+  obtain ⟨m, hm, hlt⟩ := norm_mk_lt y hδnorm,
+  have hrw : ∥y∥ + min ε (1 / 2) / (1 - min ε (1 / 2)) * ∥y∥ =
+    ∥y∥ * (1 + min ε (1 / 2) / (1 - min ε (1 / 2))) := by ring,
+  rw [hrw] at hlt,
+  have hm0 : ∥m∥ ≠ 0,
+  { intro h0,
+    have hnorm := quotient_norm_mk_le S m,
+    rw [h0, hm] at hnorm,
+    replace hnorm := le_antisymm hnorm (norm_nonneg _),
+    simpa [hnorm] using hy },
+  replace hlt := (div_lt_div_right (lt_of_le_of_ne (norm_nonneg m) hm0.symm)).2 hlt,
+  simp only [hm0, div_self, ne.def, not_false_iff] at hlt,
+  have hrw₁ : ∥y∥ * (1 + min ε (1 / 2) / (1 - min ε (1 / 2))) / ∥m∥ =
+    (∥y∥ / ∥m∥) * (1 + min ε (1 / 2) / (1 - min ε (1 / 2))) := by ring,
+  rw [hrw₁] at hlt,
+  replace hlt := (inv_pos_lt_iff_one_lt_mul (lt_trans (div_pos hδpos hδ) (lt_one_add _))).2 hlt,
+  suffices : ∥S.normed_mk∥ ≥ 1 - min ε (1 / 2),
+  { exact sub_le_iff_le_add.mp this },
+  calc ∥S.normed_mk∥ ≥ ∥(S.normed_mk) m∥ / ∥m∥ : ratio_le_op_norm S.normed_mk m
+  ...  = ∥y∥ / ∥m∥ : by rw [normed_mk.apply, hm]
+  ... ≥ (1 + min ε (1 / 2) / (1 - min ε (1 / 2)))⁻¹ : le_of_lt hlt
+  ... = 1 - min ε (1 / 2) : by field_simp [(ne_of_lt hδ).symm]
+end
+
+/-- The operator norm of the projection is `0` if the subspace is dense. -/
+lemma norm_trivial_quotient_mk (S : add_subgroup M)
+  (h : (S.topological_closure : set M) = set.univ) : ∥S.normed_mk∥ = 0 :=
+begin
+  refine le_antisymm (op_norm_le_bound _ (le_refl _) (λ x, _)) (norm_nonneg _),
+  have hker : x ∈ (S.normed_mk).ker.topological_closure,
+  { rw [S.ker_normed_mk],
+    exact set.mem_of_eq_of_mem h trivial },
+  rw [ker_normed_mk] at hker,
+  simp only [(quotient_norm_eq_zero_iff S x).mpr hker, normed_mk.apply, zero_mul],
+end
+
+/-- `is_quotient f`, for `f : M ⟶ N` means that `N` is isomorphic to the quotient of `M`
+by the kernel of `f`. -/
+structure is_quotient (f : normed_group_hom M N) : Prop :=
```

Comment:
This (and all that follows it) should be in the `normed_group_hom` namespace.

Reply:
#7653